### ISSUE TITLE
Unexpected slowdown

### ISSUE THREAD

**Issue by Callidon** [Callidon/sparql-engine#6]
_Originally posted by @dwhitney in https://github.com/Callidon/sparql-engine/issues/3#issuecomment-426502492_

> Hmm... I'm seeing some pretty significant slowdown in master now. A query that previously took roughly 400ms is now taking 40 *seconds*. Odd thing is it seems to only be happening in the browser and not in node. Here's a screenshot of some profiling I've been doing. I'll post more as I get more results.

<img width="525" alt="screen shot 2018-10-02 at 11 41 21 pm" src="https://user-images.githubusercontent.com/33441/46388924-176d3c00-c69d-11e8-9efe-29a909af26e6.png">

**Comment by Callidon** [comment]
@dwhitney I suspect that `asynciterator` is killing us here. The utility function used to wrap `Readable` into Asynciterator seems to generate deadlocks.... I was able to write a UNION query that resolved in **several minutes**, instead of a couple of milliseconds before the rework.

I will investigate a solution without using the `asynciterator` package. An idea could be to move the entire architecture in push-based, which would allow easy compatibility with Node.js Readable & Events API.

**Comment by Callidon** [comment]
@dwhitney I've pushed a complete rework on the [`push-rework`](https://github.com/Callidon/sparql-engine/tree/push-rework) branch. Could you mind giving it a try and report any remaining performance issues?

This new version completely drops `asynciterator` and now rely on push-streams. I've used the [`rxjs`](https://rxjs-dev.firebaseapp.com) package to handle the current pipeline but it is fully compatible with vanilla Javascript/Node.js structures 🎉  
Basically, `Graph.find`must now returns one of the following:
* An **array** of RDF Triples
* A [**Javascript Iterator**](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols), which yields RDF triples
* A **Readable** which yields RDF triples
* An **EventEmitter** which emits RDF triples on a `data` event

`Graph.evalBGP` has been reworked with the same requirements. I've updated the README the reflect the API evolution.

**Comment by dwhitney** [comment]
Hey Thomas, sounds great! I'm working against a deadline at the moment but
I will come back to this work in the middle of next week and I will take a
look!

Dustin

**Comment by dwhitney** [comment]
well... it took me longer to take a look than "next week", but I just did and WOWZERS! So much faster! It's already merged, but I think we can close this (I don't see an option to close on my end)

**Comment by Callidon** [comment]
Nice to hear 👍
